### PR TITLE
=lib Make use of CompletionStage#handle instead of whenComplete

### PR DESCRIPTION
**Pull request by He-Pin**
Which has better performance.

@lrytz Would you like to give some input, thanks.

Another one was :
https://github.com/line/armeria/pull/1440
https://github.com/line/armeria/pull/1444
https://github.com/akka/akka/pull/31283

I think this will help especially when the CompletionStage was failed already.

### CODE REVIEW

**Comment by SethTisue** [comment]
@viktorklang ?

**Comment by He-Pin** [comment]
image
image

jmh:run -i 5 -wi 5 -f1 -t1 .CompletionStageBenchmark.

```scala
package bench

import org.openjdk.jmh.annotations._

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@State(Scope.Benchmark)
@OutputTimeUnit(TimeUnit.SECONDS)
@BenchmarkMode(Array(Mode.Throughput))
class CompletionStageBenchmark {
  val ex = new RuntimeException("ex")

  @Benchmark
  def benchWhenCompleteFailedSync(): Unit = {
    val future = new CompletableFuture[Int]()
    future.completeExceptionally(ex)
    val latch = new CountDownLatch(1)
    future.whenComplete((_, _) => latch.countDown())
    latch.await()
  }

  @Benchmark
  def benchWhenCompleteSuccessSync(): Unit = {
    val future = new CompletableFuture[Int]()
    future.complete(1)
    val latch = new CountDownLatch(1)
    future.whenComplete((_, _) => latch.countDown())
    latch.await()
  }

  @Benchmark
  def benchHandleFailedSync(): Unit = {
    val future = new CompletableFuture[Int]()
    future.completeExceptionally(ex)
    val latch = new CountDownLatch(1)
    future.handle((_, _) => latch.countDown())
    latch.await()
  }

  @Benchmark
  def benchHandleSuccessSync(): Unit = {
    val future = new CompletableFuture[Int]()
    future.complete(1)
    val latch = new CountDownLatch(1)
    future.handle((_, _) => latch.countDown())
    latch.await()
  }

}

```

**Comment by He-Pin** [comment]
I submitted the same to kt too: https://github.com/Kotlin/kotlinx.coroutines/pull/3475

**Comment by He-Pin** [comment]
image

**Comment by He-Pin** [comment]
refs: https://github.com/scala/scala/pull/7970/commits/e3d202e09f99a22a9cfa0cfb5e4dc163e30ab5ba too

File: src/library/scala/jdk/javaapi/FutureConverters.scala [hunk 1]
```diff
@@ -72,7 +72,12 @@ object FutureConverters {
       case f: Future[T @unchecked] => f
       case _ =>
         val p = new P[T](cs)
-        cs whenComplete p
+        val future = cs.toCompletableFuture
```
Comment (lrytz) [inline src/library/scala/jdk/javaapi/FutureConverters.scala]: > A CompletionStage implementation that does not choose to interoperate with others may throw UnsupportedOperationException

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html#toCompletableFuture--

Should that case be considered?

File: src/library/scala/jdk/javaapi/FutureConverters.scala [hunk 2]
```diff
@@ -72,7 +72,12 @@ object FutureConverters {
...
+        if (future.isDone && !future.isCompletedExceptionally) {
+          p.tryComplete(Success(future.join()))
+        } else {
+          cs.handle(p)
```
Comment (lrytz) [inline src/library/scala/jdk/javaapi/FutureConverters.scala]: Do you know why `handle` is faster than `whenComplete`? Is the performance change significant?

Comment (He-Pin) [reply src/library/scala/jdk/javaapi/FutureConverters.scala]: Expecially on the failed case, I think the main outcome is from avoiding the `encodeThrowable`

File: src/library/scala/jdk/javaapi/FutureConverters.scala (hunk 1 above)
Comment (He-Pin) [reply src/library/scala/jdk/javaapi/FutureConverters.scala]: I have updated the pr to address this, thanks.

**Review by lrytz — APPROVED** [review]
Thank you @He-Pin

File: src/library/scala/jdk/javaapi/FutureConverters.scala [hunk 3]
```diff
@@ -70,9 +70,13 @@ object FutureConverters {
       case cf: CF[T] => cf.wrapped
       // in theory not safe (could be `class C extends Future[A] with CompletionStage[B]`):
       case f: Future[T @unchecked] => f
+      case cf: CompletableFuture[T @unchecked] if cf.isDone && !cf.isCompletedExceptionally =>
```
Comment (nmarton-da) [inline src/library/scala/jdk/javaapi/FutureConverters.scala:73]: FYI @He-Pin
JDK 11 has MinimalStage which is a CompletableFuture (also CompletionStage), with UnsupportedOperationException in lot of method bodies (like isDone).
With this change we throw from here at line 73 if encounered a MinimalStage.

I am not claiming this is a bug, or MinimalStage is flawed; neither know whether it is a valid case to have MinimalStage flying around, just letting you know.
Don't shoot the messenger please ;)
Others also implement safety measures: https://github.com/Kotlin/kotlinx.coroutines/issues/2456

Comment (lrytz) [reply src/library/scala/jdk/javaapi/FutureConverters.scala:73]: Many thanks @nmarton-da. It reproduces indeed:

```scala
scala> import scala.jdk.FutureConverters._
import scala.jdk.FutureConverters._

scala> import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletableFuture

scala> CompletableFuture.completedStage(42).asScala
java.lang.UnsupportedOperationException
  at java.base/java.util.concurrent.CompletableFuture$MinimalStage.isDone(CompletableFuture.java:2976)
  at scala.jdk.javaapi.FutureConverters$.asScala(FutureConverters.scala:73)
  at scala.jdk.FutureConverters$CompletionStageOps$.asScala$extension(FutureConverters.scala:41)
  ... 30 elided
```

Comment (lrytz) [reply src/library/scala/jdk/javaapi/FutureConverters.scala:73]: It seems play framework had to work around it: https://github.com/playframework/playframework/pull/11993

Comment (He-Pin) [reply src/library/scala/jdk/javaapi/FutureConverters.scala:73]: Sorry for the delay, I was busy at shopping festival,  will submit fix to other repository this weekend.

**Pull request merged**